File: frontend/apps/ios/Pantopus/Features/Me/ProfileInsightCards.swift
//
//  ProfileInsightCards.swift
//  Pantopus
//
//  Two profile-tab cards RN ships and native was missing:
//
//  * Monthly Receipt — earnings / spending / neighbors helped / posts, an
//    expandable detail block, and a share sheet. Auto-expands when the
//    profile is opened from the `monthly_receipt` push.
//    Mirrors RN `components/profile/MonthlyReceiptCard.tsx`.
//  * Invite progress — referral count against the unlock tiers, unlocked
//    feature chips, next unlock, and a share CTA carrying the invite code.
//    Mirrors RN `components/profile/InviteProgressCard.tsx`.
//

// swiftlint:disable function_parameter_count

import SwiftUI

// MARK: - Invite tiers

/// The unlock ladder, verbatim from `backend/services/inviteRewardService.js:19`.
/// The server also returns the label on `next_unlock`; this table backs the
/// milestone dots and the unlocked-feature chip labels.
public enum InviteFeatureTier: String, CaseIterable, Sendable {
    case activityMap = "activity_map"
    case neighborhoodInsights = "neighborhood_insights"
    case priorityMatching = "priority_matching"
    case foundingBadge = "founding_badge"

    public var threshold: Int {
        switch self {
        case .activityMap: 1
        case .neighborhoodInsights: 3
        case .priorityMatching: 5
        case .foundingBadge: 10
        }
    }

    public var label: String {
        switch self {
        case .activityMap: "Neighborhood Activity Map"
        case .neighborhoodInsights: "Neighborhood Insights"
        case .priorityMatching: "Priority Matching"
        case .foundingBadge: "Founding Neighbor Badge"
        }
    }

    /// Server keys we don't know fall back to the raw key rather than
    /// being dropped, so a new backend tier still renders.
    public static func label(forKey key: String) -> String {
        InviteFeatureTier(rawValue: key)?.label ?? key
    }

    public static var maxThreshold: Int {
        allCases.map(\.threshold).max() ?? 1
    }
}

// MARK: - Monthly receipt card

public struct MonthlyReceiptCard: View {
    private let receipt: MonthlyReceiptDTO
    private let onShare: @MainActor () -> Void
    @State private var isExpanded: Bool

    public init(
        receipt: MonthlyReceiptDTO,
        startExpanded: Bool = false,
        onShare: @escaping @MainActor () -> Void = {}
    ) {
        self.receipt = receipt
        self.onShare = onShare
        _isExpanded = State(initialValue: startExpanded)
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            header
            if let highlight = receipt.highlight, !highlight.isEmpty {
                Text(highlight)
                    .pantopusTextStyle(.small)
                    .italic()
                    .foregroundStyle(Theme.Color.appTextSecondary)
            }
            statsGrid
            if isExpanded { details }
            shareButton
        }
        .padding(Spacing.s4)
        .background(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .fill(Theme.Color.appSurface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .accessibilityIdentifier("monthlyReceiptCard")
    }

    private var header: some View {
        Button { isExpanded.toggle() } label: {
            HStack(spacing: Spacing.s2) {
                Icon(.barChart3, size: 18, color: Theme.Color.primary600)
                Text("\(receipt.period.label) Summary")
                    .font(.system(size: 16, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                Spacer(minLength: Spacing.s0)
                Icon(
                    isExpanded ? .chevronUp : .chevronDown,
                    size: 20,
                    color: Theme.Color.appTextMuted
                )
            }
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("monthlyReceipt.toggle")
    }

    private var statsGrid: some View {
        LazyVGrid(
            columns: [
                GridItem(.flexible(), spacing: Spacing.s2),
                GridItem(.flexible(), spacing: Spacing.s2)
            ],
            spacing: Spacing.s2
        ) {
            statBox(
                icon: .briefcase,
                tint: Theme.Color.success,
                background: Theme.Color.successBg,
                value: Self.dollars(receipt.earnings.totalCents),
                label: "Earned",
                identifier: "monthlyReceipt.earned"
            )
            statBox(
                icon: .creditCard,
                tint: Theme.Color.primary600,
                background: Theme.Color.primary50,
                value: Self.dollars(receipt.spending.totalCents),
                label: "Spent",
                identifier: "monthlyReceipt.spent"
            )
            statBox(
                icon: .users,
                tint: Theme.Color.business,
                background: Theme.Color.appSurfaceSunken,
                value: "\(receipt.community.neighborsHelped)",
                label: "Neighbors helped",
                identifier: "monthlyReceipt.neighborsHelped"
            )
            statBox(
                icon: .file,
                tint: Theme.Color.warning,
                background: Theme.Color.appSurfaceSunken,
                value: "\(receipt.community.postsCreated)",
                label: "Posts",
                identifier: "monthlyReceipt.posts"
            )
        }
    }

    private func statBox(
        icon: PantopusIcon,
        tint: Color,
        background: Color,
        value: String,
        label: String,
        identifier: String
    ) -> some View {
        VStack(spacing: Spacing.s1) {
            Icon(icon, size: 18, color: tint)
            Text(value)
                .font(.system(size: 18, weight: .bold))
                .foregroundStyle(tint)
            Text(label)
                .font(.system(size: 11))
                .foregroundStyle(Theme.Color.appTextSecondary)
                .multilineTextAlignment(.center)
        }
        .frame(maxWidth: .infinity)
        .padding(Spacing.s3)
        .background(
            RoundedRectangle(cornerRadius: Radii.md, style: .continuous).fill(background)
        )
        .accessibilityElement(children: .combine)
        .accessibilityIdentifier(identifier)
    }

    private var details: some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            Rectangle().fill(Theme.Color.appBorderSubtle).frame(height: 1)
            detailSection("Marketplace", rows: [
                ("Listings sold", "\(receipt.marketplace.listingsSold)"),
                ("Listings bought", "\(receipt.marketplace.listingsBought)"),
                ("Free items claimed", "\(receipt.marketplace.freeItemsClaimed)")
            ])
            detailSection("Earnings breakdown", rows: [
                ("Gigs completed (as worker)", "\(receipt.earnings.gigCount)"),
                ("Top category", receipt.earnings.topCategory ?? "N/A")
            ])
            detailSection("Community", rows: [
                ("Connections made", "\(receipt.community.connectionsMade)")
            ])
            detailSection("Reputation", rows: [
                ("Current rating", Self.rating(receipt.reputation.currentRating)),
                ("Reviews received", "\(receipt.reputation.reviewsReceived)"),
                ("Rating change", Self.signedRating(receipt.reputation.ratingChange))
            ])
        }
        .accessibilityIdentifier("monthlyReceipt.details")
    }

    private func detailSection(_ title: String, rows: [(String, String)]) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s1) {
            Text(title.uppercased())
                .font(.system(size: 11, weight: .semibold))
                .tracking(0.6)
                .foregroundStyle(Theme.Color.appTextMuted)
            ForEach(rows, id: \.0) { row in
                HStack {
                    Text(row.0)
                        .font(.system(size: 13.5))
                        .foregroundStyle(Theme.Color.appTextSecondary)
                    Spacer(minLength: Spacing.s2)
                    Text(row.1)
                        .font(.system(size: 13.5, weight: .semibold))
                        .foregroundStyle(Theme.Color.appText)
                }
            }
        }
    }

    private var shareButton: some View {
        Button(action: onShare) {
            HStack(spacing: Spacing.s2) {
                Icon(.share, size: 16, color: Theme.Color.primary600)
                Text("Share your month")
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundStyle(Theme.Color.primary600)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, Spacing.s3)
            .background(
                RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
                    .stroke(Theme.Color.primary600, lineWidth: 1)
            )
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("monthlyReceipt.share")
    }

    // MARK: Formatting — the server owns the numbers; we only format.

    static func dollars(_ cents: Int) -> String {
        String(format: "$%.2f", Double(cents) / 100)
    }

    static func rating(_ value: Double?) -> String {
        guard let value, value > 0 else { return "N/A" }
        return String(format: "%.1f", value)
    }

    static func signedRating(_ value: Double?) -> String {
        guard let value else { return "N/A" }
        return (value >= 0 ? "+" : "") + String(format: "%.2f", value)
    }

    /// The message RN's `handleShareReceipt` composes.
    static func shareMessage(_ receipt: MonthlyReceiptDTO) -> String {
        let earned = String(format: "%.2f", Double(receipt.earnings.totalCents) / 100)
        let highlight = receipt.highlight.map { "\($0) " } ?? ""
        return "My \(receipt.period.label) on Pantopus: \(highlight)"
            + "I earned $\(earned) and helped \(receipt.community.neighborsHelped) neighbors."
    }
}

// MARK: - Invite progress card

public struct InviteProgressCard: View {
    private let progress: InviteProgressDTO
    private let onShare: @MainActor () -> Void

    public init(progress: InviteProgressDTO, onShare: @escaping @MainActor () -> Void = {}) {
        self.progress = progress
        self.onShare = onShare
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s3) {
            Text("Invite neighbors, unlock features")
                .font(.system(size: 16, weight: .bold))
                .foregroundStyle(Theme.Color.appText)
            progressTrack
            if !progress.unlockedFeatures.isEmpty {
                unlockedChips
            }
            if let next = progress.nextUnlock {
                nextUnlockRow(next)
            }
            shareButton
        }
        .padding(Spacing.s4)
        .background(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .fill(Theme.Color.appSurface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: Radii.lg, style: .continuous)
                .stroke(Theme.Color.appBorder, lineWidth: 1)
        )
        .accessibilityIdentifier("inviteProgressCard")
    }

    private var fraction: Double {
        let maxThreshold = Double(InviteFeatureTier.maxThreshold)
        guard maxThreshold > 0 else { return 0 }
        return min(Double(progress.totalConverted) / maxThreshold, 1)
    }

    private var progressTrack: some View {
        VStack(alignment: .leading, spacing: Spacing.s2) {
            GeometryReader { geo in
                ZStack(alignment: .leading) {
                    Capsule().fill(Theme.Color.appSurfaceSunken)
                    Capsule()
                        .fill(Theme.Color.primary600)
                        .frame(width: max(0, geo.size.width * fraction))
                }
            }
            .frame(height: 6)
            HStack {
                ForEach(InviteFeatureTier.allCases, id: \.self) { tier in
                    let unlocked = progress.totalConverted >= tier.threshold
                    Text("\(tier.threshold)")
                        .font(.system(size: 10, weight: .bold))
                        .foregroundStyle(
                            unlocked ? Theme.Color.appTextInverse : Theme.Color.appTextMuted
                        )
                        .frame(width: 22, height: 22)
                        .background(
                            Circle().fill(
                                unlocked ? Theme.Color.primary600 : Theme.Color.appSurfaceSunken
                            )
                        )
                        .accessibilityIdentifier("inviteProgress.tier.\(tier.rawValue)")
                    if tier != InviteFeatureTier.allCases.last {
                        Spacer(minLength: Spacing.s1)
                    }
                }
            }
        }
        .accessibilityIdentifier("inviteProgress.track")
    }

    private var unlockedChips: some View {
        HStack(spacing: Spacing.s2) {
            ForEach(progress.unlockedFeatures, id: \.self) { key in
                HStack(spacing: Spacing.s1) {
                    Icon(.checkCircle, size: 13, color: Theme.Color.success)
                    Text(InviteFeatureTier.label(forKey: key))
                        .font(.system(size: 11.5, weight: .semibold))
                        .foregroundStyle(Theme.Color.success)
                        .lineLimit(1)
                }
                .padding(.horizontal, Spacing.s2)
                .padding(.vertical, Spacing.s1)
                .background(Capsule().fill(Theme.Color.successBg))
            }
        }
        .accessibilityIdentifier("inviteProgress.unlocked")
    }

    private func nextUnlockRow(_ next: InviteProgressDTO.NextUnlock) -> some View {
        HStack(spacing: Spacing.s2) {
            Icon(.lock, size: 13, color: Theme.Color.appTextMuted)
            Text(Self.nextUnlockText(next))
                .font(.system(size: 13))
                .foregroundStyle(Theme.Color.appTextSecondary)
            Spacer(minLength: Spacing.s0)
        }
        .accessibilityIdentifier("inviteProgress.nextUnlock")
    }

    private var shareButton: some View {
        Button(action: onShare) {
            HStack(spacing: Spacing.s2) {
                Icon(.send, size: 16, color: Theme.Color.appTextInverse)
                Text("Invite a neighbor")
                    .font(.system(size: 15, weight: .bold))
                    .foregroundStyle(Theme.Color.appTextInverse)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, Spacing.s3)
            .background(
                RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
                    .fill(Theme.Color.primary600)
            )
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("inviteProgress.share")
    }

    static func nextUnlockText(_ next: InviteProgressDTO.NextUnlock) -> String {
        let remaining = max(0, next.invitesRemaining)
        let label = next.label ?? InviteFeatureTier.label(forKey: next.feature)
        return "Invite \(remaining) more neighbor\(remaining == 1 ? "" : "s") to unlock \(label)"
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/you/me/ProfileInsightCards.kt
@file:Suppress("MagicNumber", "PackageNaming", "LongMethod", "TooManyFunctions")

package app.pantopus.android.ui.screens.you.me

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.pantopus.android.data.api.models.users.InviteNextUnlockDto
import app.pantopus.android.data.api.models.users.InviteProgressDto
import app.pantopus.android.data.api.models.users.MonthlyReceiptDto
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing
import java.util.Locale

/**
 * Two profile-tab cards RN ships and native was missing:
 *
 * * Monthly Receipt — earnings / spending / neighbors helped / posts, an
 *   expandable detail block, and a share sheet. Auto-expands when the profile
 *   is opened from the `monthly_receipt` push. Mirrors RN
 *   `components/profile/MonthlyReceiptCard.tsx`.
 * * Invite progress — referral count against the unlock tiers, unlocked
 *   feature chips, next unlock, and a share CTA carrying the invite code.
 *   Mirrors RN `components/profile/InviteProgressCard.tsx`.
 */

/**
 * The unlock ladder, verbatim from
 * `backend/services/inviteRewardService.js:19`.
 */
enum class InviteFeatureTier(
    val key: String,
    val threshold: Int,
    val label: String,
) {
    ActivityMap("activity_map", 1, "Neighborhood Activity Map"),
    NeighborhoodInsights("neighborhood_insights", 3, "Neighborhood Insights"),
    PriorityMatching("priority_matching", 5, "Priority Matching"),
    FoundingBadge("founding_badge", 10, "Founding Neighbor Badge"),
    ;

    companion object {
        val maxThreshold: Int get() = entries.maxOf { it.threshold }

        /**
         * Server keys we don't know fall back to the raw key rather than
         * being dropped, so a new backend tier still renders.
         */
        fun labelForKey(key: String): String = entries.firstOrNull { it.key == key }?.label ?: key
    }
}

/** Formatting helpers, mirrored 1:1 with iOS `MonthlyReceiptCard`. */
object MonthlyReceiptFormat {
    fun dollars(cents: Int): String = String.format(Locale.US, "\$%.2f", cents / 100.0)

    fun rating(value: Double?): String = if (value == null || value <= 0) "N/A" else String.format(Locale.US, "%.1f", value)

    fun signedRating(value: Double?): String {
        if (value == null) return "N/A"
        val sign = if (value >= 0) "+" else ""
        return sign + String.format(Locale.US, "%.2f", value)
    }

    /** The message RN's `handleShareReceipt` composes. */
    fun shareMessage(receipt: MonthlyReceiptDto): String {
        val earned = String.format(Locale.US, "%.2f", receipt.earnings.totalCents / 100.0)
        val highlight = receipt.highlight?.let { "$it " }.orEmpty()
        return "My ${receipt.period.label} on Pantopus: $highlight" +
            "I earned \$$earned and helped ${receipt.community.neighborsHelped} neighbors."
    }
}

@Composable
fun MonthlyReceiptCard(
    receipt: MonthlyReceiptDto,
    startExpanded: Boolean = false,
    onShare: () -> Unit = {},
    modifier: Modifier = Modifier,
) {
    var isExpanded by remember(startExpanded) { mutableStateOf(startExpanded) }
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .padding(Spacing.s4)
                .testTag("monthlyReceiptCard"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .clickable { isExpanded = !isExpanded }
                    .testTag("monthlyReceipt.toggle"),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            PantopusIconImage(
                icon = PantopusIcon.BarChart3,
                contentDescription = null,
                size = 18.dp,
                tint = PantopusColors.primary600,
            )
            Spacer(Modifier.width(Spacing.s2))
            Text(
                text = "${receipt.period.label} Summary",
                fontSize = 16.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appText,
                modifier = Modifier.weight(1f),
            )
            PantopusIconImage(
                icon = if (isExpanded) PantopusIcon.ChevronUp else PantopusIcon.ChevronDown,
                contentDescription = null,
                size = 20.dp,
                tint = PantopusColors.appTextMuted,
            )
        }
        receipt.highlight?.takeIf { it.isNotEmpty() }?.let {
            Text(
                text = it,
                fontSize = 14.sp,
                fontStyle = FontStyle.Italic,
                color = PantopusColors.appTextSecondary,
            )
        }
        ReceiptStatsGrid(receipt)
        if (isExpanded) ReceiptDetails(receipt)
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(Radii.md))
                    .border(1.dp, PantopusColors.primary600, RoundedCornerShape(Radii.md))
                    .clickable(onClick = onShare)
                    .padding(vertical = Spacing.s3)
                    .testTag("monthlyReceipt.share"),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            PantopusIconImage(
                icon = PantopusIcon.Share,
                contentDescription = null,
                size = 16.dp,
                tint = PantopusColors.primary600,
            )
            Spacer(Modifier.width(Spacing.s2))
            Text(
                text = "Share your month",
                fontSize = 14.sp,
                fontWeight = FontWeight.SemiBold,
                color = PantopusColors.primary600,
            )
        }
    }
}

@Composable
private fun ReceiptStatsGrid(receipt: MonthlyReceiptDto) {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.s2)) {
        Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            StatBox(
                icon = PantopusIcon.Briefcase,
                tint = PantopusColors.success,
                background = PantopusColors.successBg,
                value = MonthlyReceiptFormat.dollars(receipt.earnings.totalCents),
                label = "Earned",
                tag = "monthlyReceipt.earned",
                modifier = Modifier.weight(1f),
            )
            StatBox(
                icon = PantopusIcon.CreditCard,
                tint = PantopusColors.primary600,
                background = PantopusColors.primary50,
                value = MonthlyReceiptFormat.dollars(receipt.spending.totalCents),
                label = "Spent",
                tag = "monthlyReceipt.spent",
                modifier = Modifier.weight(1f),
            )
        }
        Row(horizontalArrangement = Arrangement.spacedBy(Spacing.s2)) {
            StatBox(
                icon = PantopusIcon.Users,
                tint = PantopusColors.business,
                background = PantopusColors.appSurfaceSunken,
                value = "${receipt.community.neighborsHelped}",
                label = "Neighbors helped",
                tag = "monthlyReceipt.neighborsHelped",
                modifier = Modifier.weight(1f),
            )
            StatBox(
                icon = PantopusIcon.File,
                tint = PantopusColors.warning,
                background = PantopusColors.appSurfaceSunken,
                value = "${receipt.community.postsCreated}",
                label = "Posts",
                tag = "monthlyReceipt.posts",
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun StatBox(
    icon: PantopusIcon,
    tint: Color,
    background: Color,
    value: String,
    label: String,
    tag: String,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .clip(RoundedCornerShape(Radii.md))
                .background(background)
                .padding(Spacing.s3)
                .testTag(tag),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(icon = icon, contentDescription = null, size = 18.dp, tint = tint)
        Text(text = value, fontSize = 18.sp, fontWeight = FontWeight.Bold, color = tint)
        Text(
            text = label,
            fontSize = 11.sp,
            color = PantopusColors.appTextSecondary,
            textAlign = TextAlign.Center,
        )
    }
}

@Composable
private fun ReceiptDetails(receipt: MonthlyReceiptDto) {
    Column(
        modifier = Modifier.fillMaxWidth().testTag("monthlyReceipt.details"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Box(Modifier.fillMaxWidth().height(1.dp).background(PantopusColors.appBorderSubtle))
        DetailSection(
            "Marketplace",
            listOf(
                "Listings sold" to "${receipt.marketplace.listingsSold}",
                "Listings bought" to "${receipt.marketplace.listingsBought}",
                "Free items claimed" to "${receipt.marketplace.freeItemsClaimed}",
            ),
        )
        DetailSection(
            "Earnings breakdown",
            listOf(
                "Gigs completed (as worker)" to "${receipt.earnings.gigCount}",
                "Top category" to (receipt.earnings.topCategory ?: "N/A"),
            ),
        )
        DetailSection(
            "Community",
            listOf("Connections made" to "${receipt.community.connectionsMade}"),
        )
        DetailSection(
            "Reputation",
            listOf(
                "Current rating" to MonthlyReceiptFormat.rating(receipt.reputation.currentRating),
                "Reviews received" to "${receipt.reputation.reviewsReceived}",
                "Rating change" to MonthlyReceiptFormat.signedRating(receipt.reputation.ratingChange),
            ),
        )
    }
}

@Composable
private fun DetailSection(
    title: String,
    rows: List<Pair<String, String>>,
) {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.s1)) {
        Text(
            text = title.uppercase(Locale.US),
            fontSize = 11.sp,
            fontWeight = FontWeight.SemiBold,
            letterSpacing = 0.6.sp,
            color = PantopusColors.appTextMuted,
        )
        rows.forEach { (label, value) ->
            Row(modifier = Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = label,
                    fontSize = 13.5.sp,
                    color = PantopusColors.appTextSecondary,
                    modifier = Modifier.weight(1f),
                )
                Text(
                    text = value,
                    fontSize = 13.5.sp,
                    fontWeight = FontWeight.SemiBold,
                    color = PantopusColors.appText,
                )
            }
        }
    }
}

@Composable
fun InviteProgressCard(
    progress: InviteProgressDto,
    onShare: () -> Unit = {},
    modifier: Modifier = Modifier,
) {
    Column(
        modifier =
            modifier
                .fillMaxWidth()
                .clip(RoundedCornerShape(Radii.lg))
                .background(PantopusColors.appSurface)
                .border(1.dp, PantopusColors.appBorder, RoundedCornerShape(Radii.lg))
                .padding(Spacing.s4)
                .testTag("inviteProgressCard"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Text(
            text = "Invite neighbors, unlock features",
            fontSize = 16.sp,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
        )
        InviteProgressTrack(progress.totalConverted)
        if (progress.unlockedFeatures.isNotEmpty()) {
            Row(
                modifier = Modifier.fillMaxWidth().testTag("inviteProgress.unlocked"),
                horizontalArrangement = Arrangement.spacedBy(Spacing.s2),
            ) {
                progress.unlockedFeatures.forEach { key ->
                    Row(
                        modifier =
                            Modifier
                                .clip(RoundedCornerShape(Radii.pill))
                                .background(PantopusColors.successBg)
                                .padding(horizontal = Spacing.s2, vertical = Spacing.s1),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        PantopusIconImage(
                            icon = PantopusIcon.CheckCircle,
                            contentDescription = null,
                            size = 13.dp,
                            tint = PantopusColors.success,
                        )
                        Spacer(Modifier.width(Spacing.s1))
                        Text(
                            text = InviteFeatureTier.labelForKey(key),
                            fontSize = 11.5.sp,
                            fontWeight = FontWeight.SemiBold,
                            color = PantopusColors.success,
                            maxLines = 1,
                        )
                    }
                }
            }
        }
        progress.nextUnlock?.let { next ->
            Row(
                modifier = Modifier.fillMaxWidth().testTag("inviteProgress.nextUnlock"),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                PantopusIconImage(
                    icon = PantopusIcon.Lock,
                    contentDescription = null,
                    size = 13.dp,
                    tint = PantopusColors.appTextMuted,
                )
                Spacer(Modifier.width(Spacing.s2))
                Text(
                    text = nextUnlockText(next),
                    fontSize = 13.sp,
                    color = PantopusColors.appTextSecondary,
                )
            }
        }
        Row(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(Radii.md))
                    .background(PantopusColors.primary600)
                    .clickable(onClick = onShare)
                    .padding(vertical = Spacing.s3)
                    .testTag("inviteProgress.share"),
            horizontalArrangement = Arrangement.Center,
            verticalAlignment = Alignment.CenterVertically,
        ) {
            PantopusIconImage(
                icon = PantopusIcon.Send,
                contentDescription = null,
                size = 16.dp,
                tint = PantopusColors.appTextInverse,
            )
            Spacer(Modifier.width(Spacing.s2))
            Text(
                text = "Invite a neighbor",
                fontSize = 15.sp,
                fontWeight = FontWeight.Bold,
                color = PantopusColors.appTextInverse,
            )
        }
    }
}

@Composable
private fun InviteProgressTrack(totalConverted: Int) {
    val maxThreshold = InviteFeatureTier.maxThreshold
    val fraction = (totalConverted.toFloat() / maxThreshold).coerceIn(0f, 1f)
    Column(
        modifier = Modifier.fillMaxWidth().testTag("inviteProgress.track"),
        verticalArrangement = Arrangement.spacedBy(Spacing.s2),
    ) {
        Box(
            modifier =
                Modifier
                    .fillMaxWidth()
                    .height(6.dp)
                    .clip(RoundedCornerShape(Radii.pill))
                    .background(PantopusColors.appSurfaceSunken),
        ) {
            Box(
                modifier =
                    Modifier
                        .fillMaxWidth(fraction)
                        .height(6.dp)
                        .clip(RoundedCornerShape(Radii.pill))
                        .background(PantopusColors.primary600),
            )
        }
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            InviteFeatureTier.entries.forEach { tier ->
                val unlocked = totalConverted >= tier.threshold
                Box(
                    modifier =
                        Modifier
                            .size(22.dp)
                            .clip(CircleShape)
                            .background(
                                if (unlocked) PantopusColors.primary600 else PantopusColors.appSurfaceSunken,
                            ).testTag("inviteProgress.tier.${tier.key}"),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        text = "${tier.threshold}",
                        fontSize = 10.sp,
                        fontWeight = FontWeight.Bold,
                        color = if (unlocked) PantopusColors.appTextInverse else PantopusColors.appTextMuted,
                    )
                }
            }
        }
    }
}

/** System share sheet for the receipt / invite CTAs. */
internal fun shareText(
    context: android.content.Context,
    message: String,
) {
    val send =
        android.content.Intent(android.content.Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(android.content.Intent.EXTRA_TEXT, message)
        }
    context.startActivity(android.content.Intent.createChooser(send, "Share"))
}

internal fun nextUnlockText(next: InviteNextUnlockDto): String {
    val remaining = next.invitesRemaining.coerceAtLeast(0)
    val label = next.label ?: InviteFeatureTier.labelForKey(next.feature)
    return "Invite $remaining more neighbor${if (remaining == 1) "" else "s"} to unlock $label"
}
